Change get_devices function, complete tests, make fixes

webdjoe opened the pull request:
Prevent get_devices from instantiating all new devices. Finish tests for all devices (89% coverage), make associated fixes.  Adjust travis.yml to run pytest.

Add __str__, __repr__, __eq__ and __hash__ methods to devices.  __hash__ used CID attribute and __eq__ uses CID to compare devices.

webdjoe committed:
Update tests and related fixes

webdjoe committed:
flake8 linting and travis.yml update

webdjoe committed:
More travis fixes

webdjoe committed:
Remove unittest line from travis

pytest also runs unittests

webdjoe committed:
Fix typo in get_details in air purifier object

webdjoe committed:
Add uuid to request body

markperdue commented on src/pyvesync/helpers.py:
```diff
@@ -169,29 +169,56 @@ def call_api(api: str, method: str,
+                if all(x in resp for x in keys):
+                    return True
+                else:
+                    logger.error('Keys missing from get_details for 7A')
```
should there be a return here to match all the other branches?

webdjoe replied:
Yes, definitely.  Nice catch

markperdue commented on src/pyvesync/helpers.py:
```diff
@@ -169,29 +169,56 @@ def call_api(api: str, method: str,
-                return True
             else:
                 return False
+            logger.error('Unkown call')
```
would this ever be triggered?

webdjoe replied:
No this is a formatting error.  The `logger.error` line should be indented above the `return False`.  The idea is if we are adding a device we can trace where the error occurs.  I fixed the formatting.  I'll take it out if you think it is redundant.

markperdue commented on src/pyvesync/vesync.py:
```diff
@@ -72,22 +73,84 @@ def energy_update_interval(self, new_energy_update):
             self._energy_update_interval = new_energy_update
 
+    @staticmethod
+    def device_search(device, new_list):
```
`device_search` makes it sound more like this method would return whether a device exists as opposed to what the docstring says this method does. what do you think on renaming this to something more clearer?

webdjoe replied:
Agreed, renamed to `remove_dev_test`

markperdue commented on src/pyvesync/vesync.py:
```diff
@@ -72,22 +73,84 @@ def energy_update_interval(self, new_energy_update):
+            else:
+                return True
+
+    def new_devices(self, new_dev):
```
same as the other one. maybe something like `should_add_device` or something else?

webdjoe replied:
I agree, renamed to `add_dev_test` for consistency (the sun was coming up when I finished this off and I forgot to revisit the names the next day)

markperdue commented on src/pyvesync/vesync.py:
```diff
@@ -98,17 +161,18 @@ def process_devices(self, devices) -> tuple:
                 #    bulbs.append(VSFactory.getDevice(devType, dev, self))
                 else:
-                    logger.debug('Unknown device ' + devType)
+                    logger.error('Unknown device ' + devType)
```
is there a reason to switch off of debug to error? error statements are logged in HA by default so we might be adding noise to the logs unless there is a good reason

webdjoe replied:
The logic behind this is if someone buys a new device and it's not supported, it will show up in the logs to indicate why the device isn't appearing.  Basically to skip a step in troubleshooting if someone raises an issue with an unsupported device.  I was on the fence with this.  What do you think?

markperdue commented on src/pyvesync/vesyncbasedevice.py:
```diff
@@ -11,55 +12,59 @@ class VeSyncBaseDevice(object):
 
-    def is_on(self) -> str:
+    def __repr__(self):
+        return "DevClass: {}, Name:{}, DevStatus: {}, CID: {}" \
```
What is DevClass?

webdjoe replied:
`__repr__` is only used for consoles and debugging to give a string representation of devices.  The `DevClass` shows the class the device was instantiated on.  The `__repr__` function allows the object to return a unique readable string if inside of a list.  I used it to debug the device lists in the vesync object when I was working on the get_devices function and tests.

markperdue commented on src/pyvesync/vesyncoutlet.py:
```diff
@@ -105,84 +105,78 @@ def yearly_energy_total(self) -> float:
-            'get',
-            headers=helpers.req_headers(self.manager)
-        )
+        r, _ = helpers.call_api('/v1/device/' + self.cid + '/detail',
```
why all these changes? the previous form is still valid on a lot of these changes and all the changes are making the PR huge

webdjoe replied:
I agree, I reverted all of the changes in your following comments and made the call_api formatting consistent throughout the code.  I'll try to keep the PR's smaller in the future

markperdue commented on src/pyvesync/vesyncoutlet.py:
```diff
@@ -232,67 +224,60 @@ def get_weekly_energy(self):
-            headers=helpers.req_headers(self.manager),
-            json=body
-        )
+        response, _ = helpers.call_api('/10a/v1/device/energyweek',
```
this refactor makes it harder to read

markperdue commented on src/pyvesync/vesyncoutlet.py:
```diff
@@ -232,67 +224,60 @@ def get_weekly_energy(self):
-            headers=helpers.req_headers(self.manager),
-            json=body
-        )
+        response, _ = helpers.call_api('/10a/v1/device/energymonth',
```
harder to read with refactor

webdjoe committed:
Address comments

Fixed get_devices function names to be more discriptive
Adjusted formating of call_api functions
Fixed unreachable logger function in helpers.py
Added logger functions in fan and switch

webdjoe committed:
Additional formatting fixed

webdjoe committed:
Fix test error check

webdjoe committed:
Fix display function, fix dev list update

Fix error in display function, fix device list append to extend

webdjoe committed:
Standardize logging levels for devices

webdjoe committed:
Fix air purifier change_fan_speed method

markperdue commented:
Changes look good. thanks!
